modules/internal/OwnedObject.chpl

```diff
@@ -432,7 +432,14 @@ module OwnedObject {
   // Don't print out 'chpl_p' when printing an _owned, just print class pointer
   pragma "no doc"
   proc _owned.readWriteThis(f) throws {
-    f <~> this.chpl_p;
+    if isNonNilableClass(this.chpl_t) {
+      var tmp = this.chpl_p!;
+      f <~> tmp;
+      if tmp == nil then halt("internal error - read nil");
```

Comment:
Couldn't this happen because of non-internal reasons? Like when reading from a corrupt channel or something? If so, we can be more verbose here. e.g. "read nil for a non-nilable class type" or something

Reply:
Nope, the I/O code doesn't try to read `nil` for the class in that event.

I added a test showing the behavior when reading `nil` into a non-nilable class and made some related improvements.